lib/module_player.js

```diff
@@ -0,0 +1,213 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+import {delay, delayThenReject} from './promise.js';
+
+export function configure({makeEmptyModule, monitor, debug, time, logError}) {
+  function logIfError(fn) {
+    try {
+      fn();
+    } catch (e) {
+      logError(e);
+    }
+  }
+
+  return class ModulePlayer {
+    static makeEmptyModule() {
+      return makeEmptyModule();
+    }
+    constructor() {
+      // start with some kind of initial, off module.
+      this.oldModule = ModulePlayer.makeEmptyModule();
+      // When next module is non-null, it means that we've got a request come in
+      // to play a new module.
+      this.nextModule = null;
+      // When true, we are actively trying to transition to a new module (i.e,
+      // goToModule is in-progress).
+      this.transitionInProgress = false;
+    }
+    // Behavior we want:
+    // - No more than one goToModule operation is happening at a time.
+    // - If playModule is called only once, goToModule completes.
+    // - If playModule is called while goToModule is happening, goToModule aborts
+    //   early at known interruption points, and we go to the new module.
+    // - If playModule is called more than once while goToModule is happening,
+    //   only the most recent becomes the nextModule.
+    // You only need to wait for playModule if you want to be sure that the
+    // player has tried to transition to your module, which is perfect for
+    // tests.
+    async playModule(module) {
+      this.nextModule = module;
+      if (monitor.isEnabled()) {
+        monitor.update({
+          event: `playModule: ${module.name}`,
+          time: time.now(),
+          deadline: module.deadline
+        });
+      }
+      await this.performTransitions();
+    }
+
+    async performTransitions() {
+      if (this.transitionInProgress) {
+        return; // We'll do this work on the next iteration of the loop below.
+      }
+      this.transitionInProgress = true;
+      while (this.oldModule != this.nextModule) {
+        const module = this.nextModule;
+        // Our local copy of the module we are trying to go to.
+        await this.goToModule(this.nextModule);
+        // Maybe we succeeded in going to module... maybe we didn't. It isn't
+        // clear. If we failed to go to the module, then we should stop trying
+        // to do so, but only if we haven't been told to go somewhere else.
+        const failedToTransition = module != this.oldModule;
```

Comment:
There's a bad smell around both this function and goToModule depending so much on this.oldModule and this.nextModule being updated "just so." Not sure what to do about it, feels like it needs a whiteboard.

Reply:
I was trying to do this without introducing new state... any specific suggestions for improvement?